Add memory format support to `cpu` and `cuda` operators

VitalyFedyunin opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #28077 Add channels last support to cuda.comm.scatter and gather
* #28076 Kill `operator==` of TensorOptions as confusing one
* #27979 Add memory format support to `resize_as_` operator
* #27891 Cleanup testing of _like operators
* #27890 Add memory format support to `randn_like` operator
* #27889 Add memory format support to `randint_like` operator
* #27562 Add memory format support to `zeros_like` operator
* #27561 Add memory format support to `rand_like` operator
* #27270 Add memory format support to `ones_like` operator
* #27262 Add memory format support to `full_like` operator
* #27244 Add memory format support to `empty_like` operator
* #27228 Add memory format support to typecasting shortcuts `byte`,`char`,`double`,`bool`,`half`,`int`,`long`,`short`,`float`,`bfloat16`
* **#27223 Add memory format support to `cpu` and `cuda` operators**

Adds memory_format keyword argument (positional for cpp).

'Preserve' behavior now follows next rules:
1) If tensor is non-overlapping and dense - output tensor will have the same strides as input tensor.
2) If not (1) and tensor is stored in the channels last format, output tensor going to have channels last format.
3) Output tensor is going to be contiguous in all other cases.

---
Dense tensor is the tensor that store values in a contiguous block of memory.
Non-overlapping tensor is the tensor in which elements occupy individual non-repetitive memory.

Differential Revision: [D17980128](https://our.internmc.facebook.com/intern/diff/D17980128)

VitalyFedyunin committed:
Add memory format support to `cpu` and `cuda` operators

[ghstack-poisoned]

VitalyFedyunin committed:
Update on "Add memory format support to `cpu` and `cuda` operators"




[ghstack-poisoned]

VitalyFedyunin committed:
Update on "Add memory format support to `cpu` and `cuda` operators"




[ghstack-poisoned]

VitalyFedyunin committed:
Update on "Add memory format support to `cpu` and `cuda` operators"


Adds memory_format keyword argument (positional for cpp).

'Preserve' behavior now follows next rules:
1) If tensor is non-overlapping and dense - output tensor will have the same strides as input tensor.
2) If not (1) and tensor is stored in the channels last format, output tensor going to have channels last format.
3) Output tensor is going to be contiguous in all other cases.

---
Dense tensor is the tensor that store values in a contiguous block of memory.
Non-overlapping tensor is the tensor in which elements occupy individual non-repetitive memory.

[ghstack-poisoned]

VitalyFedyunin committed:
Update on "Add memory format support to `cpu` and `cuda` operators"


Adds memory_format keyword argument (positional for cpp).

'Preserve' behavior now follows next rules:
1) If tensor is non-overlapping and dense - output tensor will have the same strides as input tensor.
2) If not (1) and tensor is stored in the channels last format, output tensor going to have channels last format.
3) Output tensor is going to be contiguous in all other cases.

---
Dense tensor is the tensor that store values in a contiguous block of memory.
Non-overlapping tensor is the tensor in which elements occupy individual non-repetitive memory.

[ghstack-poisoned]

VitalyFedyunin committed:
Update on "Add memory format support to `cpu` and `cuda` operators"


Adds memory_format keyword argument (positional for cpp).

'Preserve' behavior now follows next rules:
1) If tensor is non-overlapping and dense - output tensor will have the same strides as input tensor.
2) If not (1) and tensor is stored in the channels last format, output tensor going to have channels last format.
3) Output tensor is going to be contiguous in all other cases.

---
Dense tensor is the tensor that store values in a contiguous block of memory.
Non-overlapping tensor is the tensor in which elements occupy individual non-repetitive memory.

[ghstack-poisoned]

VitalyFedyunin committed:
Update on "Add memory format support to `cpu` and `cuda` operators"


Adds memory_format keyword argument (positional for cpp).

'Preserve' behavior now follows next rules:
1) If tensor is non-overlapping and dense - output tensor will have the same strides as input tensor.
2) If not (1) and tensor is stored in the channels last format, output tensor going to have channels last format.
3) Output tensor is going to be contiguous in all other cases.

---
Dense tensor is the tensor that store values in a contiguous block of memory.
Non-overlapping tensor is the tensor in which elements occupy individual non-repetitive memory.

[ghstack-poisoned]

VitalyFedyunin committed:
Fix merge conflict on "Add memory format support to `cpu` and `cuda` operators"


Adds memory_format keyword argument (positional for cpp).

'Preserve' behavior now follows next rules:
1) If tensor is non-overlapping and dense - output tensor will have the same strides as input tensor.
2) If not (1) and tensor is stored in the channels last format, output tensor going to have channels last format.
3) Output tensor is going to be contiguous in all other cases.

---
Dense tensor is the tensor that store values in a contiguous block of memory.
Non-overlapping tensor is the tensor in which elements occupy individual non-repetitive memory.

[ghstack-poisoned]

VitalyFedyunin committed:
Fix .to copy=True path on "Add memory format support to `cpu` and `cuda` operators"


Adds memory_format keyword argument (positional for cpp).

'Preserve' behavior now follows next rules:
1) If tensor is non-overlapping and dense - output tensor will have the same strides as input tensor.
2) If not (1) and tensor is stored in the channels last format, output tensor going to have channels last format.
3) Output tensor is going to be contiguous in all other cases.

---
Dense tensor is the tensor that store values in a contiguous block of memory.
Non-overlapping tensor is the tensor in which elements occupy individual non-repetitive memory.

[ghstack-poisoned]

VitalyFedyunin commented:
@ailzhang breaks XLA

gchanan reviewed (approved):
lgtm.

It would be nice if there were JIT tests for all of this.  This area of the code is janky enough that I could believe: 1) JIT doesn't support it, 2) JIT tries to support it, but it's not consistent with the eager implementation 3) JIT does support it, but the support is fragile and we could break it without intending to.

VitalyFedyunin committed:
Rebase on top of master on "Add memory format support to `cpu` and `cuda` operators"


Adds memory_format keyword argument (positional for cpp).

'Preserve' behavior now follows next rules:
1) If tensor is non-overlapping and dense - output tensor will have the same strides as input tensor.
2) If not (1) and tensor is stored in the channels last format, output tensor going to have channels last format.
3) Output tensor is going to be contiguous in all other cases.

---
Dense tensor is the tensor that store values in a contiguous block of memory.
Non-overlapping tensor is the tensor in which elements occupy individual non-repetitive memory.

[ghstack-poisoned]

VitalyFedyunin committed:
Rebasing to master on "Add memory format support to `cpu` and `cuda` operators"


Adds memory_format keyword argument (positional for cpp).

'Preserve' behavior now follows next rules:
1) If tensor is non-overlapping and dense - output tensor will have the same strides as input tensor.
2) If not (1) and tensor is stored in the channels last format, output tensor going to have channels last format.
3) Output tensor is going to be contiguous in all other cases.

---
Dense tensor is the tensor that store values in a contiguous block of memory.
Non-overlapping tensor is the tensor in which elements occupy individual non-repetitive memory.

Differential Revision: [D17980128](https://our.internmc.facebook.com/intern/diff/D17980128)

[ghstack-poisoned]
